dvc/objects/db/base.py

```diff
@@ -53,7 +53,10 @@ def add(self, path_info, fs, hash_info, **kwargs):
         # using our makedirs to create dirs with proper permissions
         self.makedirs(cache_info.parent)
         if isinstance(fs, type(self.fs)):
-            self.fs.move(path_info, cache_info)
+            if move:
+                self.fs.move(path_info, cache_info)
+            else:
+                self.fs.copy(path_info, cache_info)
```

Comment:
This one in preparation for https://github.com/iterative/dvc/issues/5811 , right? Maybe let's not add it for now and just fallback to `else: upload_fobj` that we have below.

Reply:
No, it is still needed. It removes the data from the original location if both the source and the remote storage have the same fs type.

Reply:
@isidentical I mean the `fs.copy` part. It could be:
```
if isinstance(fs, type(self.fs)) and move:
    self.fs.move()
else:
     with fs.open()...
         self.fs.upload_fobj()
```

or am I missing something?


On this topic, seems like we are finally reaching the point where we actually need to distinguish move from rename and copy from serverside-copy, so that we could provide the most efficient experience and nicest code. But it will do for now.

Reply:
That also works!